server/src/main/java/org/cloudfoundry/identity/uaa/user/JdbcUaaUserDatabase.java

```diff
@@ -253,20 +285,23 @@ protected void getAuthorities(Set<String> authorities, final List<String> member
                 Object[] parameterList = ArrayUtils.addAll(new Object[] { identityZoneManager.getCurrentIdentityZoneId() }, memberList.subList(0, size).toArray());
 
                 results.addAll(jdbcTemplate.queryForList(dynamicAuthoritiesQuery.toString(), parameterList));
-
                 memberList = memberList.subList(size, memberList.size());
             }
-            List<String> newMemberIdList = new ArrayList<>();
+            return results;
+        }
 
-            for (Map<String, Object> record : results) {
-                String displayName = (String) record.get("displayName");
-                String groupId = (String) record.get("id");
-                if (!authorities.contains(displayName)) {
-                    authorities.add(displayName);
-                    newMemberIdList.add(groupId);
-                }
-            }
-            getAuthorities(authorities, newMemberIdList);
+        private List<Map<String, Object>> executeAuthoritiesQueryPostgresql(List<String> memberList) {
+            String arrayAuthoritiesQuery = "select g.id,g.displayName from groups g, group_membership m where g.id = m.group_id  and g.identity_zone_id=? and m.member_id = ANY(?)";
```

Comment:
The only comment here I have is.
Did you executed this via explain ? because of issue https://github.com/cloudfoundry/uaa/issues/1662 ?
I recommend to check for indexed query plan here and then also think about of LOWER for identity zone id

Reply:
Checked the statement via explain analyze:
the ANY(?) is using group_membership_unique_key to do an Index Only Scan
the groups id join is using groups_pkey to do an Index Scan and then filtering on identity_zone_id = 'uaa' (so no scan over identity zone id is happening here)

Reply:
ok, thanks